### DISCUSSION TITLE
How to watch for most recent value, then stop? ie: ref.watchOnce()

### DISCUSSION THREAD

**Discussion by oodavid** [rrousselGit/riverpod#1069] [Ideas]
Like many on Flutter, we're using Firestore for our backend.

By default we have document / collection streams, ie:

```dart
final docProvider = StreamProvider.autoDispose<Doc>((ref) {
  /// firestore....snapshots() here
});
```

Most of the time we want realtime behaviour:

```dart
// This will rebuild every time the document / stream is updated
Widget build(BuildContext context, WidgetRef ref) {
  AsyncValue<Doc> doc = ref.watch(docProvider);
  return doc.when(
    loading: () => const CircularProgressIndicator(),
    error: (error, stack) => const Text('Oops'),
    data: (doc) => Text(doc.contents),
  );
}
```

Occasionally we'd just like the most recent value, and don't want to rebuild on subsequent yields.

```dart
// This will build data once with a loading / error state:
Widget build(BuildContext context, WidgetRef ref) {
  AsyncValue<Doc> doc = ref.watchOnce(docProvider);
  return doc.when(
    loading: () => const CircularProgressIndicator(),
    error: (error, stack) => const Text('Oops'),
    data: (doc) => Text(doc.contents),
  );
}
```

The above example uses streams, but the question would apply to computed providers too, ie:

```dart
final modifiedDocProvider = Provider.autoDispose<ModifiedDoc>((ref) {
  AsyncValue<Doc> doc = ref.watch(docProvider);
  return ModifiedDoc.fromDoc(doc);
});

// using it:
AsyncValue<ModifiedDoc> modifiedDoc = ref.watchOnce(modifiedDocProvider);
```

We'd benefit for not having to duplicate our stream logic for equivalents with futures. I think this would be more pronounced with the computed providers.

**Comment by rrousselGit** [comment]
Ignoring new values is anti-pattern.

What's the goal?

**Reply by oodavid** [reply]
Happy to defer to your experience here TBH.

Without going into details, the goal is to have a "view" screen that is realtime, and an "edit" screen that ignores realtime updates, but still wants to access that most recent state during initialisation.

**Reply by rrousselGit** [reply]
You can use "ref.listen" instead of "ref.watch" inside your computed provider if you to manually handle updates. This should allow you to "nothing" when the value changes
